File: 이것이 취업을 위한 코딩 테스트다/그래프_이론/크루스칼/크루스칼.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class 크루스칼 {
  public static class Edge implements Comparable<Edge> {
    public int distance;
    public int nodeA;
    public int nodeB;

    public Edge(int distance, int nodeA, int nodeB) {
      this.distance = distance;
      this.nodeA = nodeA;
      this.nodeB = nodeB;
    }

    @Override
    public int compareTo(Edge o) {
      if (this.distance < o.distance) {
        return -1;
      }

      return 1;
    }
  }

  public static int v, e;
  public static int[] parent = new int[10000001];
  public static ArrayList<Edge> edges = new ArrayList<>();
  public static int result = 0;

  public static int findParent(int x) {
    if (x != parent[x]) {
      parent[x] = findParent(parent[x]);
    }

    return parent[x];
  }

  public static void unionParent(int a, int b) {
    a = findParent(a);
    b = findParent(b);

    if (a < b) {
      parent[b] = a;
    } else {
      parent[a] = b;
    }
  }

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    v = sc.nextInt();
    e = sc.nextInt();

    for (int i = 1; i <= v; i++) {
      parent[i] = i;
    }

    for (int i = 0; i < e; i++) {
      int a = sc.nextInt();
      int b = sc.nextInt();
      int cost = sc.nextInt();
      edges.add(new Edge(cost, a, b));
    }

    Collections.sort(edges);

    for (int i = 0; i < edges.size(); i++) {
      int cost = edges.get(i).distance;
      int a = edges.get(i).nodeA;
      int b = edges.get(i).nodeB;

      if (findParent(a) != findParent(b)) {
        unionParent(a, b);
        result += cost;
      }
    }
    System.out.println(result);
  }
}

File: 이것이 취업을 위한 코딩 테스트다/그래프_이론/크루스칼/크루스칼.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int v, e;
int parent[100001];
vector<pair<int, pair<int, int>>> edges;
int result = 0;

int findParent(int x) {
  if (x != parent[x]) {
    parent[x] = findParent(parent[x]);
  }

  return parent[x];
}

void unionParent(int a, int b) {
  a = findParent(a);
  b = findParent(b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
}

int main(void) {
  cin >> v >> e;

  for (int i = 1; i <= v; i++) {
    parent[i] = i;
  }

  for (int i = 0; i < e; i++) {
    int a, b, c;
    cin >> a >> b >> c;

    edges.push_back({c, {a, b}});
  }

  sort(edges.begin(), edges.end());

  for(int i = 0; i < edges.size(); i++) {
    int c = edges[i].first;
    int a = edges[i].second.first;
    int b = edges[i].second.second;

    if (findParent(a) != findParent(b)) {
      unionParent(a, b);
      result += c;
    }
  }

  cout << result;
}